test/precompile.jl

```diff
@@ -1554,5 +1554,35 @@ precompile_test_harness("issue #46296") do load_path
     (@eval (using CodeInstancePrecompile))
 end
 
+precompile_test_harness("foreign types") do path
+    write(joinpath(load_path, "Foreign.jl"),
+        """
+        module Foreign
+
+        export FObj
+
+        const FObj = ccall(:jl_new_foreign_type, Any, (Symbol, Module, Any, Any, Any, Cint, Cint),
+                        :FObj, Foreign, Any, C_NULL, C_NULL, 0, 0)
+
+        FObj() = ccall(:jl_new_struct_uninit, Any, (Any,), FObj)
+
+        const nmark = Ref(0)
+        const nsweep = Ref(0)
+        inc_nmark() = (nmark[] += 1; return nothing)
+        inc_nsweep() = (nsweep[] += 1; return nothing)
+
+        function __init__()
+            ccall(:jl_reinit_foreign_type, Cint, (Any, Any, Any),
+                FObj, @cfunction(inc_nmark, Cvoid, ()), @cfunction(inc_nsweep, Cvoid, ()))
+        end
+
+        end # module Foreign
+        """)
+    (@eval (using Foreign))
+    x = [FObj() for _ in 1:1000]
```

Comment:
@timholy wrote elsewhere:

> Not quite sure why `nmark` and `nsweep` are still unincremented after calling `GC.gc(true)`, is that a sign that something still isn't working?

Is perhaps `x` keeping the references alive? Perhaps just try this:
```suggestion
    [FObj() for _ in 1:1000]
```

One could also use a classic finalizer to test if the object gets finalized. I.e., insert something like this 
```suggestion
    x = [FObj() for _ in 1:1000]
    x_was_finalized = false
    finalizer(x) do x
        x_was_finalized = true
    end
```
and then check `x_was_finalized` is set to `true`.

Or combine both ideas:
```suggestion
    x_was_finalized = false
    finalizer([FObj() for _ in 1:1000]) do x
        x_was_finalized = true
    end
```